### PR TITLE
Pcap Directory Mode (Feature #2222) v8

### PR DESCRIPTION
**Pull request by dbcfd**
Version 8 of 
 - https://github.com/OISF/suricata/pull/2985
 - https://github.com/OISF/suricata/pull/2971
 - https://github.com/OISF/suricata/pull/2964
 - https://github.com/OISF/suricata/pull/2961
 - https://github.com/OISF/suricata/pull/2960
 - https://github.com/OISF/suricata/pull/2958
 - https://github.com/OISF/suricata/pull/2957

Pcap file mode that when passed a directory will process all files in that directory. If continuous flag is set (--pcap-file-continuous), directory will be monitored until the directory is moved/deleted or suricata is interrupted.

Engine state will not reset between files.

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2222

Describe changes:
- If source-pcap-file is passed a directory instead of a file, all files in that directory will be processed in order of modified time, without state resetting between files.
- source-pcap-file will continue to process files until suricata is interrupted or the directory is moved/deleted when using --pcap-file-continuous or continuous json argument.
- runmode-unix-socket is updated to provide a current time to determine what state of processing of the directory is

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: src/source-pcap-file.c [hunk 1]
```diff
@@ -44,6 +44,7 @@
 #include "runmode-unix-socket.h"
```
Comment (victorjulien) [inline src/source-pcap-file.c]: think we can split this file? I think it's getting a bit hard to follow all the logic. Perhaps a src/source-pcap-file-dir or src/source-pcap-dir?

**Comment by victorjulien** [comment]
My CentOS 7 builder is unhappy:
<pre>
source-pcap-file.c: In function ‘ReceivePcapFileLoop’:
source-pcap-file.c:206:11: error: ‘modified_time.tv_nsec’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
         if(left->tv_nsec < right->tv_nsec) {
           ^
source-pcap-file.c:469:29: note: ‘modified_time.tv_nsec’ was declared here
             struct timespec modified_time;
                             ^
source-pcap-file.c:203:14: error: ‘modified_time.tv_sec’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
     } else if(left->tv_sec > right->tv_sec) {
              ^
source-pcap-file.c:469:29: note: ‘modified_time.tv_sec’ was declared here
             struct timespec modified_time;
                             ^
source-pcap-file.c:197:17: error: ‘temp_time.tv_nsec’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
     to->tv_nsec = from->tv_nsec;
                 ^
source-pcap-file.c:832:37: note: ‘temp_time.tv_nsec’ was declared here
                     struct timespec temp_time;
                                     ^
source-pcap-file.c:196:16: error: ‘temp_time.tv_sec’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
     to->tv_sec = from->tv_sec;
                ^
source-pcap-file.c:832:37: note: ‘temp_time.tv_sec’ was declared here
                     struct timespec temp_time;
                                     ^
cc1: all warnings being treated as errors
</pre>

**Comment by victorjulien** [comment]
It fails on many others of my builders too.

File: src/source-pcap-file.c [hunk 2]
```diff
@@ -60,47 +61,519 @@
...
+                else if (CompareTimes(&temp_time, older_than) >= 0) {
+                    SCLogDebug("Skipping new file %s", pathbuff);
+                    continue;
+                }
+            } else {
+                SCLogDebug("Unable to get modified time on %s, skipping", pathbuff);
+                continue;
+            }
+
+            file_to_add = SCMalloc(sizeof(PendingFile));
```
Comment (victorjulien) [inline src/source-pcap-file.c]: prefer SCCalloc as that inits mem to 0

File: src/source-pcap-file.c [hunk 3]
```diff
@@ -60,47 +61,519 @@
...
+            SCLogDebug("Found \"%s\"", file_to_add->filename);
+
+            if(PcapDirectoryInsertFile(pv, file_to_add, &temp_time) == TM_ECODE_FAILED) {
+                SCLogError(SC_ERR_INVALID_ARGUMENT, "Failed to add file");
+                SCReturnInt(TM_ECODE_FAILED);
+            }
+        }
+        else
+        {
+            SCLogError(SC_ERR_SPRINTF, "Could not write path");
```
Comment (victorjulien) [inline src/source-pcap-file.c]: think it's nicer to check this right after the snprintf call

File: src/runmode-unix-socket.c [hunk 4]
```diff
@@ -310,41 +390,73 @@ static TmEcode UnixSocketPcapFilesCheck(void *data)
...
+    if(set_res != 1) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Can not set continuous");
+        PcapFilesFree(cfile);
+        return TM_ECODE_FAILED;
+    }
+
+    if(cfile->delay > 0) {
+        char tstr[32];
+        snprintf(tstr, sizeof(tstr), "%ld", cfile->delay);
+        if (ConfSetFinal("pcap-file.delay", tstr) != 1) {
```
[the remark this answers is not in the document]
Comment (jasonish) [reply src/runmode-unix-socket.c]: Probably not. Never thought of the use case of setting values after initialization. Probably best avoided.

[the remark this answers is not in the document]
Comment (victorjulien) [reply src/runmode-unix-socket.c]: I think the existing code does some of this as well. It's also one of the reasons I did the 'prefix' loading for tenants. Then at least it's adding stuff in a new namespace.

I don't think there is a lot of code that does Conf calls after startup, so perhaps we can just wrap a lock around things.

Another problem is the potential for consumers of the api to keep reference to data in the api. If that does away we'll end up with dangling pointers. Ideally the api should allow for it.

[the remark this answers is not in the document]
Comment (victorjulien) [reply src/runmode-unix-socket.c]: Edit: "Ideally the api should NOT allow for it"

File: src/source-pcap-file.c [hunk 5]
```diff
@@ -60,47 +61,519 @@
...
+                }
+                file_to_compare = next_file_to_compare;
+            }
+        }
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+TmEcode PcapDirectoryPopulateBuffer(
```
[the remark this answers is not in the document]
Comment (dbcfd) [reply src/source-pcap-file.c]: Yes, tailq

File: src/source-pcap-file.c (hunk 1 above)
Comment (dbcfd) [reply src/source-pcap-file.c]: I'm looking at leaving source-pcap-file in place focusing on the receive/decode thread functionality, and introducing pcap-file-helper and pcap-directory-helper with the implementation related to that functionality.

Comment (victorjulien) [reply src/source-pcap-file.c]: yeah makes sense

File: src/runmode-unix-socket.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (dbcfd) [reply src/runmode-unix-socket.c]: Yeah the existing code is doing the same conf settings https://github.com/OISF/suricata/blob/master/src/runmode-unix-socket.c#L316.

It's sort of kind of okay since the threads are then getting kicked off to read the values. 

Is it ok for this PR to keep this in place, or is there an alternative?

[the remark this answers is not in the document]
Comment (victorjulien) [reply src/runmode-unix-socket.c]: I guess we can let it go in, as it should be reasonably safe. But if you're up for it, I would prefer a thread safe solution. Something like a mutex protected TAILQ that we add jobs to so we wouldn't need to touch the Conf API at all.

[the remark this answers is not in the document]
Comment (dbcfd) [reply src/runmode-unix-socket.c]: Due to the size of the PR, created https://redmine.openinfosecfoundation.org/issues/2296 to look into a way to not touch the Conf API.

**Comment by dbcfd** [comment]
Closing for https://github.com/OISF/suricata/pull/3027
